quil-rs/src/reserved.rs

```diff
@@ -20,79 +25,36 @@ impl FromStr for ReservedToken {
     type Err = NotReservedToken;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        if let Ok(keyword) = ReservedKeyword::from_str(s) {
-            Ok(Self::Keyword(keyword))
-        } else if let Ok(gate) = ReservedGate::from_str(s) {
-            Ok(Self::Gate(gate))
-        } else if let Ok(constant) = ReservedConstant::from_str(s) {
-            Ok(Self::Constant(constant))
-        } else {
-            Err(NotReservedToken(s.to_string()))
+        fn parse<T: FromStr>(
+            reserved: impl Fn(T) -> ReservedToken,
```

Comment:
Can the `Self` type be used here? Not saying it would be better, just curious the scoping

```suggestion
            reserved: impl Fn(T) -> Self,
```

Reply:
It actually can’t – I wrote it that way first, but you get “error[E0401]: can't use `Self` from outer item”.